shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/Rules/TestChain.hs

```diff
@@ -600,6 +602,28 @@ preserveOutputsTx SourceSignalTarget {source = chainSt, signal = block} =
        in property $
             hasFailedScripts tx || outs `Map.isSubmapOf` u'
 
+canRestrictUTxO ::
+  forall era ledger.
+  ( ChainProperty era,
+    TestingLedger era ledger,
+    HasField "inputs" (Core.TxBody era) (Set (TxIn (Crypto era)))
+  ) =>
+  SourceSignalTarget (CHAIN era) ->
+  Property
+canRestrictUTxO SourceSignalTarget {source = chainSt, signal = block} =
+  conjoin $
+    map outputPreserved $
+      zip (sourceSignalTargets ledgerTrFull) (sourceSignalTargets ledgerTrRestr)
+  where
+    (_, ledgerTrFull) = ledgerTraceFromBlock @era @ledger chainSt block
+    (UTxO irrelevantUTxO, ledgerTrRestr) =
+      ledgerTraceFromBlockWithRestrictedUTxO @era @ledger chainSt block
+    outputPreserved
+      ( SourceSignalTarget {target = (UTxOState {_utxo = (UTxO uFull)}, _), signal = _},
+        SourceSignalTarget {target = (UTxOState {_utxo = (UTxO uRestr)}, _), signal = _}
+        ) =
+        uRestr `Map.disjoint` irrelevantUTxO && uFull == uRestr `Map.union` irrelevantUTxO
```

Comment:
I always try to use QuickChecks functions, since then we'll get the output printed if it fails:

```suggestion
        (uRestr `Map.disjoint` irrelevantUTxO) .&&. uFull === (uRestr `Map.union` irrelevantUTxO)
```